File: python/codeeval/007_prefix_expressions.py
import sys

OPS = ['+', '/', '*']

def pref(line):
    s = line.split(' ')
    r = []
    for i in reversed(s):
        if i in OPS:
            r1 = r.pop()
            r2 = r.pop()
            if i == '+':
                r.append(r1 + r2)
            elif i == '*':
                r.append(r1 * r2)
            elif i == '/':
                r.append(r1 / r2)
        else:
            r.append(int(i))
    rf = r[0]
    ri = int(r[0])
    return ri if ri == rf else rf


with open(sys.argv[1], 'r') as test_cases:
    for test in test_cases:
        print(pref(test))

File: ruby/codeeval/007_prefix_expressions.rb
OPS = ['+', '/', '*']

def pref(line)
    s = line.split(' ')
    r = []
    s.reverse.each do |i|
        if OPS.include?(i)
            r1 = r.pop()
            r2 = r.pop()
            if i == '+'
                r.push(r1 + r2)
            elsif i == '*'
                r.push(r1 * r2)
            elsif i == '/'
                r.push(r1.to_f / r2.to_f)
            end
        else
            r.push(i.to_i)
        end
    end
    rf = r[0]
    ri = r[0].to_i
    return ri == rf ? ri : rf
end

File.open(ARGV[0]).each_line do |line|
    puts(pref(line))
end
